# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [17]:
import pandas as pd # 'pd' is an alias for pandas

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"

customer_data = pd.read_csv(url)

# Identify the dimensions of the dataset by determining the number of rows and columns it contains.
print("Number of rows and colums: ", customer_data.shape)



Number of rows and colums:  (4008, 11)


In [18]:
# Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
print("Types of columns: ", customer_data.dtypes)


Types of columns:  Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object


Customer Lifetime Value (object):
This should likely be a numerical column (probably a float) since it represents a monetary value. 

Gender (object):
This column is likely categorical, but it is stored as object. This might indicate there are non-numeric entries 

In [19]:
#Identify the number of unique values for each column and determine which columns appear to be categorical. 
print(customer_data.nunique())

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64


In [20]:
# You should also describe the unique values of each categorical column 
customer_data.describe(include='object')

Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   

       Number of Open Complaints    Policy Type  Vehicle Class  
count                       1071           1071           1071  
unique                         6              3              6  
top                       1/0/00  Personal Auto  Four-Door Car  
freq                         830            780            576

T (8 unique values):
Likely represents states.

GENDER (5 unique values):
The values represent different genders. However, having 5 unique values might indicate the presence of variations

Education (6 unique values):
Likely represents different education levels. 

Number of Open Complaints (6 unique values):
Likely represents the count of open complaints. The small number of unique values indicates it could be treated as categorical

Policy Type (3 unique values):
Likely represents different policy types

Vehicle Class (6 unique values):
Represents different classes of vehicles .

In [21]:
# The range of values for numerical columns, and give your insights.
customer_data.describe()

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

Income
- Range: 0 to 99960 -> Significant variability
- Mean(39295) & median 50%(36234):  -> They are close. Some customers have more or less income than the average
- Min: 0 -> Indicates some customer have no incomes

Monthly Premium Auto
- Range: 61 to 35354 -> Significant variability
- Mean(193,234360) & median 50%(83) -> Mostly customers pay lower premium. Some pay very high 
- Min: 61 -> 

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [22]:
#Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

customer_data_ST = pd.DataFrame(customer_data)
customer_data_ST = customer_data_ST["ST"].value_counts().tail(5)
customer_data_ST


ST
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [23]:
# Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

customer_data_PT = customer_data["Policy Type"].value_counts().head(1)
print(customer_data_PT.index[0])

Personal Auto


In [25]:
customer_data_PT1 = customer_data["Policy Type"].value_counts()
print(customer_data_PT1.idxmax())

Personal Auto


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [27]:
customer_data.sample()

Customer   ST GENDER Education Customer Lifetime Value  Income  \
1325      NaN  NaN    NaN       NaN                     NaN     NaN   

      Monthly Premium Auto Number of Open Complaints Policy Type  \
1325                   NaN                       NaN         NaN   

     Vehicle Class  Total Claim Amount  
1325           NaN                 NaN

In [28]:
# Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.

customer_data_PA = customer_data.loc[customer_data["Policy Type"] == "Personal Auto"]
customer_data_CA = customer_data.loc[customer_data["Policy Type"] == "Corporate Auto"]

# Calculate the average income for each policy.

personal_auto = customer_data_PA["Income"].mean()
corporate_auto = customer_data_CA["Income"].mean()


# Print the results.   

print(f'Average income for Personal Auto policies: {personal_auto}')
print(f'Average income for Corporate Auto policies: {corporate_auto}') 


Average income for Personal Auto policies: 38180.69871794872
Average income for Corporate Auto policies: 41390.31196581197


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [30]:
# Review again the statistics for total claim amount to gain an understanding of the data.Your code here

total_claim_amount = customer_data["Total Claim Amount"]
total_claim_amount.describe()

count    1071.000000
mean      404.986909
std       293.027260
min         0.382107
25%       202.157702
50%       354.729129
75%       532.800000
max      2893.239678
Name: Total Claim Amount, dtype: float64

In [35]:
# Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.

customer_data_sorted = customer_data.sort_values(by="Total Claim Amount", ascending=True)

percentile = int(0.75 * len(customer_data_sorted))

print(customer_data_sorted.iloc[percentile:])


     Customer   ST GENDER Education Customer Lifetime Value  Income  \
3006      NaN  NaN    NaN       NaN                     NaN     NaN   
3007      NaN  NaN    NaN       NaN                     NaN     NaN   
3008      NaN  NaN    NaN       NaN                     NaN     NaN   
3009      NaN  NaN    NaN       NaN                     NaN     NaN   
3010      NaN  NaN    NaN       NaN                     NaN     NaN   
...       ...  ...    ...       ...                     ...     ...   
4003      NaN  NaN    NaN       NaN                     NaN     NaN   
4004      NaN  NaN    NaN       NaN                     NaN     NaN   
4005      NaN  NaN    NaN       NaN                     NaN     NaN   
4006      NaN  NaN    NaN       NaN                     NaN     NaN   
4007      NaN  NaN    NaN       NaN                     NaN     NaN   

      Monthly Premium Auto Number of Open Complaints Policy Type  \
3006                   NaN                       NaN         NaN   
3007       

In [31]:
customer_data_high_claim = customer_data[customer_data["Total Claim Amount"] > customer_data["Total Claim Amount"].quantile(0.75)]
customer_data_high_claim

Customer          ST GENDER Education Customer Lifetime Value   Income  \
1     QZ44356     Arizona      F  Bachelor              697953.59%      0.0   
2     AI49188      Nevada      F  Bachelor             1288743.17%  48767.0   
17    OE15005        Cali    NaN   College              394524.16%  28855.0   
23    TZ98966      Nevada    NaN  Bachelor              245019.10%      0.0   
26    US89481  California    NaN  Bachelor              394637.21%      0.0   
...       ...         ...    ...       ...                     ...      ...   
1059  YG44474      Oregon      M   College             1401472.13%  54193.0   
1061  RY92647        Cali      F  Bachelor             1050677.17%      0.0   
1068  GS98873     Arizona      F  Bachelor              323912.47%  16061.0   
1069  CW49887  California      F    Master              462680.11%  79487.0   
1070  MY31220  California      F   College              899704.02%  54230.0   

      Monthly Premium Auto Number of Open Complaints     Policy Type  \
1                     94.0                    1/0/00   Personal Auto   
2                    108.0                    1/0/00   Personal Auto   
17                   101.0                    1/0/00   Personal Auto   
23                    73.0                    1/3/00  Corporate Auto   
26                   111.0                    1/0/00   Personal Auto   
...                    ...                       ...             ...   
1059                 117.0                    1/0/00  Corporate Auto   
1061                  92.0                    1/0/00   Personal Auto   
1068                  88.0                    1/0/00   Personal Auto   
1069                 114.0                    1/0/00    Special Auto   
1070                 112.0                    1/0/00   Personal Auto   

      Vehicle Class  Total Claim Amount  
1     Four-Door Car         1131.464935  
2      Two-Door Car          566.472247  
17              SUV          647.442031  
23    Four-Door Car          554.376763  
26    Four-Door Car          799.200000  
...             ...                 ...  
1059            SUV          720.752945  
1061  Four-Door Car          546.524896  
1068  Four-Door Car          633.600000  
1069            SUV          547.200000  
1070   Two-Door Car          537.600000  

[264 rows x 11 columns]